src/p11_key.c

```diff
@@ -332,6 +326,82 @@ int pkcs11_generate_key(PKCS11_SLOT_private *slot, int algorithm, unsigned int b
 	return 0;
 }
 
+/**
+ * Generate EC key pair directly on token
+ */
+int pkcs11_ec_keygen(PKCS11_SLOT_private *slot, const char *curve,
+		const char *label, const unsigned char *id, size_t id_len,
+		const PKCS11_params* params)
+{
+	PKCS11_CTX_private *ctx = slot->ctx;
+	CK_SESSION_HANDLE session;
+	PKCS11_TEMPLATE pubtmpl = {0}, privtmpl = {0};
+	CK_MECHANISM mechanism = {
+		CKM_EC_KEY_PAIR_GEN, NULL_PTR, 0
+	};
+	CK_OBJECT_HANDLE pub_key_obj, priv_key_obj;
+	int rv;
+	unsigned char *ec_params = NULL;
+	int ec_params_len = 0;
+	unsigned char *tmp = NULL;
+	ASN1_OBJECT *curve_obj = NULL;
+	int curve_nid = NID_undef;
+
+	if (pkcs11_init_keygen(slot, &session))
+		return -1;
+
+	curve_nid = EC_curve_nist2nid(curve);
+	if (curve_nid == NID_undef)
+		curve_nid = OBJ_sn2nid(curve);
+	if (curve_nid == NID_undef)
+		curve_nid = OBJ_ln2nid(curve);
+	if (curve_nid == NID_undef)
+		return -1;
+	curve_obj = OBJ_nid2obj(curve_nid);
+	if (!curve_obj)
+		return -1;
+	// convert to DER format and take just the length
```

Comment:
Use C-style comments instead of C++. This note applies to all comments; sorry for the confusion.